internal/server/singleprocess/service_trigger_test.go

```diff
@@ -382,3 +382,188 @@ func TestServiceTrigger_DeleteTrigger(t *testing.T) {
 		require.NotNil(resp)
 	})
 }
+
+func TestServiceTrigger_RunTrigger(t *testing.T) {
+	ctx := context.Background()
+
+	// Create our server
+	impl, err := New(WithDB(testDB(t)))
+	require.NoError(t, err)
+	client := server.TestServer(t, impl)
+
+	t.Run("running a missing trigger", func(t *testing.T) {
+		require := require.New(t)
+
+		// try to run a trigger, require error
+		resp, err := client.RunTrigger(ctx, &pb.RunTriggerRequest{
+			Ref: &pb.Ref_Trigger{
+				Id: "nope",
+			},
+		})
+
+		require.Error(err)
+		require.Nil(resp)
+	})
+
+	t.Run("running a trigger when project is missing", func(t *testing.T) {
+		require := require.New(t)
+
+		respTrigger, err := client.UpsertTrigger(ctx, &pb.UpsertTriggerRequest{
+			Trigger: &pb.Trigger{
+				Id:   "test",
+				Name: "build-trigger",
+				Operation: &pb.Trigger_Build{
+					Build: &pb.Job_BuildOp{
+						DisablePush: false,
+					},
+				},
+				Project: &pb.Ref_Project{
+					Project: "secret_project",
+				},
+				Application: &pb.Ref_Application{
+					Application: "another_one",
+					Project:     "secret_project",
+				},
+				Workspace: &pb.Ref_Workspace{
+					Workspace: "staging",
+				},
+			},
+		})
+		require.NoError(err)
+		require.NotNil(respTrigger)
+
+		// try to run a trigger, require error
+		resp, err := client.RunTrigger(ctx, &pb.RunTriggerRequest{
+			Ref: &pb.Ref_Trigger{
+				Id: respTrigger.Trigger.Id,
+			},
+		})
+
+		require.Error(err)
+		require.Nil(resp)
+		st, ok := status.FromError(err)
+		require.True(ok)
+		require.Equal(codes.NotFound, st.Code())
+	})
```

Comment:
In the test scenario here we create a trigger (ID generated, URLs supported, etc) for a Project that doesn't exist (yet, or possibly never will at all). So we create a valid Trigger (no error when creating), but it's also an invalid Trigger (error when trying to use). 

This scenario surprised me so I went looking. It seems [UpsertTrigger](https://github.com/hashicorp/waypoint/blob/109efacee9a7d8b4506de4ebf8e537181b03e0d6/internal/server/singleprocess/service_trigger.go#L11) doesn't validate that the `Project` referenced actually exists, just that it's not an empty string. 

This may be beyond your PR, but I feel like we should ensure it exists before we create more things based off of it, especially when those things can manipulate the thing that doesn't yet exist.

Reply:
Yep, so this behavior is intentional. I didn't require a project existing on upseting a trigger because the idea is these triggers are going to be configured ahead of time to be used later in CI. Guarding on a project existing to create a trigger would make that much more difficult to achieve. If you try to run a trigger though, we will check if it exists, which I think is the right time to check.

One middle ground would be to display a warning in the CLI if a trigger has been created but no project exists for the requested trigger.

Reply:
For future work, have we thought about adding a "Test trigger" ability? This could be CLI and a nice fancy UI button. It would cover the "OK I made a trigger; now the dev made a project; now the release engineer needs to test the trigger points to the right project."

Reply:
Maybe like a `waypoint trigger validate <id>`? It could look up the project and optionally the app to make sure they exist?